Rubocop enforcement and simplecov

stephengroat opened the pull request:
i know it's huge, if you'd like to minimize the rule impact let me know!

`rubocop` now runs with `rake` as the default

most corrections were done with `rubocop --auto-correct`, some was ignored in `.rubocop.yml`. if there's a specific style change you'd like to enforce, let me know!

gshutler commented:
Thanks for this. I've had a skim and there's a few things that stand out:

 1. The "prefer `''` over `""`" is responsible for 99% of the noise, could we suppress that one (at least temporarily) and then we might see the wood for the trees?
 2. Very wary of changing constants from `PascalCase` to `SNAKE_CASE` as people could rely on them as they're part of the public API (sort of)
 3. Similar goes for method names, would either leave as they are or alias-and-warn the old name
 4. I saw a big change to avoid `def self.foo` in `lib/user_agent/operating_systems.rb` which I think changes the accessibility of the `WINDOWS` constant, I'm fine with `def self.foo` myself
 5. I would be ok with keeping the `old => hash` style, or at least making that change it's own commit a bit like the string quotation for the sake of clarity in commit history/review

stephengroat commented:
Let me see what i can do with `.rubocop.yml`

stephengroat commented:
@gshutler See below

1. the `Style/StringLiterals#double_quotes` lessened the noise, but there's still enough difference that it creates some (since it enforces consistency)
2. added aliases with warnings for any constants that were changed to `SCREAMING_SNAKE_CASE` and tested them
3. only found one method name that `rubocop` wanted changed (`has_wmfsdk?` to `wmfsdk?`), added an alias and warned
4. moved `WINDOWS` outside of the `class` context of the module, can access now with `UserAgent::OperatingSystems::WINDOWS` or `UserAgent::OperatingSystems.Windows`
5. Updated with enforcing `hash_rockets`, didn't help as much as I had hoped

gshutler commented on lib/user_agent/browsers/internet_explorer.rb:
```diff
@@ -6,27 +6,27 @@ class InternetExplorer < Base
         "Trident/7.0" => "11.0",
         "Trident/6.0" => "10.0",
         "Trident/5.0" => "9.0",
-        "Trident/4.0" => "8.0",
```
I would allow or even enforce trailing commas as it makes for clearer diffs in the future when a new value is appended to an array like this.

gshutler replied:
It's a hash in this case, either way trailing commas are a good thing for both arrays and hashes in my experience.

stephengroat replied:
addressed for both hashes and arrays

gshutler commented on lib/user_agent/browsers/chrome.rb:
```diff
@@ -2,17 +2,22 @@ class UserAgent
+      ].freeze
+
+      def ChromeBrowsers # rubocop:disable Naming/MethodName
+        warn("#{__method__} is deprecated. Please use CHROME_BROWSERS instead")
```
Like this but I've had my logs filled by a deprecation notice like this on a hot path. Particularly annoying when it's a dependency of a dependency or something like that and so troublesome to change.

Once we've got a final Rubocop setup I'd like to change however many we have of these to something like:

```ruby
unless @chrome_browser_warned
  warn("#{__method__} is deprecated. Please use CHROME_BROWSERS instead")
  @chrome_browser_warned = true
end
```

stephengroat replied:
addressed

gshutler commented on lib/user_agent/operating_systems.rb:
```diff
@@ -18,42 +20,44 @@ module OperatingSystems
+        "OS X #{version}"
       end
+    end
+    end
```
Indentation is a bit weird in this class, starts at `class << self` I think. Surprised Rubocop didn't flag this!

stephengroat replied:
with how "verbose" it likes to be, i'm shocked too, but if rubocop is happy, i'm just happy it's not yelling

stephengroat commented:
think i addressed everything you're looking for (`user_agent_spec` now has tests showing warn once and never again for a instance, could be for a class with `@@` if you want it to be)

stephengroat commented:
@gshutler i'm sure you're super busy, i just want to make sure that this pops back up on your radar (if there's anything else you need from me)

stephengroat commented:
any interest in this? trying to cleanup my open PRs and wondering if i should just close this. if so, no worries
